enzyme/Enzyme/EnzymeLogic.cpp

```diff
@@ -3410,19 +3410,21 @@ Function *EnzymeLogic::CreatePrimalAndGradient(
           replacedReturns[orig] = si;
         }
 
-        if (dretAlloca && !gutils->isConstantValue(orig->getReturnValue())) {
-          rb.CreateStore(gutils->invertPointerM(orig->getReturnValue(), rb),
+        if (key.retType == DIFFE_TYPE::DUP_ARG || 
+	    key.retType == DIFFE_TYPE::DUP_NONEED) {
+        	if (dretAlloca) {
+          		rb.CreateStore(gutils->invertPointerM(orig->getReturnValue(), rb),
                          dretAlloca);
-        }
-        if (key.retType == DIFFE_TYPE::OUT_DIFF) {
+		}
+        } else if (key.retType == DIFFE_TYPE::OUT_DIFF) {
           assert(orig->getReturnValue());
           assert(differetval);
           if (!gutils->isConstantValue(orig->getReturnValue())) {
             IRBuilder<> reverseB(gutils->reverseBlocks[BB].back());
-            gutils->setDiffe(orig->getReturnValue(), differetval, reverseB);
+            gutils->setDiffe(orig->getReturnValue(), Constant::getNullValue(differetval->getType()), reverseB);
```

Comment:
``
```suggestion
      gutils->setDiffe(orig->getReturnValue(), differetval, reverseB);
    }
      else
            gutils->setDiffe(orig->getReturnValue(), Constant::getNullValue(differetval->getType()), reverseB);
```

Reply:
The issue with that is that it fails 9 of our tests since orig->getReturnValue() will give us a ConstantValue, which triggers an assertion in setDiffe:
https://github.com/wsmoses/Enzyme/blob/1f6f58d5bc1c1c78f0a1cd8aae2e89e9f6f17036/enzyme/Enzyme/GradientUtils.h#L1919
I didn't found out yet what's the best way to fix that, do you have a recommendation?

Reply:
Oh sorry, I thought we were setting the return gradient, not the gradient of the return value. In which case the original `gutils->setDiffe(orig->getReturnValue(), differetval, reverseB);` is fine.